### PR TITLE
packetbeat: add support for NFS v3 and v4 protocols

### PR DESCRIPTION
**Pull request by kofemann**
This pull request is more an RFC. We will run it bit in production go get more experience. Comments are welcome. The example output:
```json
{
  "@timestamp": "2016-03-28T06:18:18.431Z",
  "beat": {
    "hostname": "localhost",
    "name": "localhost"
  },
  "count": 1,
  "dst": "127.0.0.1",
  "dst_port": 2049,
  "nfs": {
    "minor_version": 1,
    "opcode": "GETATTR",
    "status": "NFSERR_NOENT",
    "tag": "",
    "version": 4
  },
  "rpc": {
    "auth_flavor": "unix",
    "call_size": 200,
    "cred": {
      "gid": 500,
      "gids": [
        491,
        499,
        500
      ],
      "machinename": "localhost",
      "stamp": 4597002,
      "uid": 500
    },
    "reply_size": 96,
    "status": "success",
    "time": 25631000,
    "time_str": "25.631ms",
    "xid": "2cf0c876"
  },
  "src": "127.0.0.1",
  "src_port": 975,
  "type": "nfs"
}
```

### CODE REVIEW

**Comment by kofemann** [comment]
The one of the questions I would like to know: the other beats use field names with dots, but elasticsearch 2.0 does not allow that. So is it ok in beats?

**Comment by kofemann** [comment]
Ok. Now it's in the shape, that  others can use it as well, IOW - this is a real pull request now.

**Comment by kofemann** [comment]
Dashboard example
screenshot from 2016-03-28 17-17-26

File: packetbeat/protos/rpc/nfs.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"github.com/elastic/beats/libbeat/common"
+	"time"
+)
+
+type Nfs struct {
+	xdr   Xdr
+	vers  uint32
+	proc  uint32
+	event *common.MapStr
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/nfs.go]: I would recommend not using a pointer here. Maps are reference types so they are always passed by reference.

File: packetbeat/protos/rpc/nfs3.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"READDIR",
+	"READDIRPLUS",
+	"FSSTAT",
+	"FSINFO",
+	"PATHINFO",
+	"COMMIT",
+}
+
+func (nfs *Nfs) getV3Opcode() string {
+	if nfs.proc < 22 {
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/nfs3.go]: Should `22` be `len(nfs_opnum3)`?

File: packetbeat/protos/rpc/nfs4.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
+package rpc
+
+import "fmt"
+
+const OP_ACCESS = 3
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/nfs4.go]: Consider grouping these constants in a single `const` block. It's common to group related constants this way.

```
const (
    OP_ACCESS = 3
    OP_CLOSE = 4
)
```

I think you could also use [`iota`](https://golang.org/ref/spec#Iota) here.

```
const (
    OP_ACCESS = iota + 3
    OP_CLOSE
    OP_COMMIT
    ...
    OP_ILLEGAL = 10044
)
```

File: packetbeat/protos/rpc/rpc.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		size := int(marker & RPC_SIZE_MASK)
+		is_last := (marker & RPC_LAST_FRAG) != 0
+
+		if len(st.rawData)-4 < size {
+			logp.Debug("rpc", "Wainting for more data")
+			break
+		}
+
+		if !is_last {
+			logp.Warn("rpc", "multifragment rpc message")
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/rpc.go]: `logp.Warn` doesn't accept a debug selector argument from what I recall.

File: packetbeat/protos/rpc/rpc.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+	st := conn.Streams[dir]
+	if st == nil {
+		st = newStream(pkt, tcptuple)
+		conn.Streams[dir] = st
+	} else {
+		// concatenate bytes
+		st.rawData = append(st.rawData, pkt.Payload...)
+		if len(st.rawData) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("rpc", "Stream data too large, dropping TCP stream")
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/rpc.go]: I would recommend creating a single `debugf` function for your package. And everywhere you write `logp.Debug("rpc", "some message")` replace it with `debugf("some message")`. It makes it easy to see what the debug selectors are for the package and it avoids the problem of typos with the selector.

Basically at the package level just add this:

`var debugf = logp.MakeDebug("rpc")`

File: packetbeat/protos/rpc/rpc_message.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	"prog_unavail",
+	"prog_mismatch",
+	"proc_unavail",
+	"garbage_args",
+	"system_err",
+}
+
+var calls_seen = common.NewCache(1*time.Minute, 8192)
+
+func (msg *RpcMessage) fillEvent(event *common.MapStr, results publish.Transactions, size int) {
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/rpc_message.go]: No need to use a MapStr pointer as mentioned earlier.

File: packetbeat/protos/rpc/rpc_message.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	if msg_type == RPC_CALL {
+
+		// eat rpc version number
+		msg.xdr.getUInt()
+
+		rpc_prog := msg.xdr.getUInt()
+		rpc_prog_vers := msg.xdr.getUInt()
+		rpc_proc := msg.xdr.getUInt()
+
+		if rpc_prog == 100003 {
```
Comment (andrewkroh) [inline packetbeat/protos/rpc/rpc_message.go]: Seems like `100003` should be named constant?

**Comment by andrewkroh** [comment]
Thanks for the contribution! Very cool screen shot. A couple miscellaneous points:

- Try running your package against `goimports`. It should change the ordering on some of the imports.
- Some of the variable names are using underscores. The Go convention is [mixedCase](https://golang.org/doc/effective_go.html#mixed-caps). Try running `golint` against your package.
- I see you added some pcaps. To ensure that there are no regressions we like to have some system tests in-place that replay the pcaps and check the outputs. Could you add some test cases like [these ones](https://github.com/elastic/beats/blob/master/packetbeat/tests/system/test_0032_dns.py) that use those pcaps?
- And finally, to make sure that everyone understands what the fields mean it would be great if you could document their meanings. We put each field into [fields.yml](https://github.com/elastic/beats/blob/master/packetbeat/etc/fields.yml) then we generate documentation from it by running `make update`.

File: packetbeat/protos/rpc/nfs.go (hunk 1 above)
Comment (kofemann) [inline packetbeat/protos/rpc/nfs.go]: fixed.

File: packetbeat/protos/rpc/nfs3.go (hunk 2 above)
Comment (kofemann) [inline packetbeat/protos/rpc/nfs3.go]: fixed

File: packetbeat/protos/rpc/nfs4.go (hunk 3 above)
Comment (kofemann) [inline packetbeat/protos/rpc/nfs4.go]: fixed

File: packetbeat/protos/rpc/rpc.go (hunk 4 above)
Comment (kofemann) [inline packetbeat/protos/rpc/rpc.go]: fixed

File: packetbeat/protos/rpc/rpc.go (hunk 5 above)
Comment (kofemann) [inline packetbeat/protos/rpc/rpc.go]: fixed

File: packetbeat/protos/rpc/rpc_message.go (hunk 6 above)
Comment (kofemann) [inline packetbeat/protos/rpc/rpc_message.go]: this was not clear to me. Thanks.

File: packetbeat/protos/rpc/rpc_message.go (hunk 7 above)
Comment (kofemann) [inline packetbeat/protos/rpc/rpc_message.go]: fixed

**Comment by kofemann** [comment]
I have addresses all issues you pointed out.. With style it was not clear, a if code used mixed style in various places, like  tcptuple.Dst_ip.String().  Any way, Tests added, rebased to current master,  branch updated.

**Comment by andrewkroh** [comment]
> The one of the questions I would like to know: the other beats use field names with dots, but elasticsearch 2.0 does not allow that. So is it ok in beats?

Because field names containing dots are not allowed in ES 2.0 we do not use them in Beats. We use nested maps which gives the appearance of field names with dots because you can use dot-notation to Kibana and ES to reference the fields. For example:

```
event = common.MapStr{}
nfs = common.MapStr{}
event["nfs"] = nfs
nfs["version"] = 4
```

File: packetbeat/tests/system/test_0061_nfs.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        )
+        self.run_packetbeat(pcap="nfs_v3.pcap")
+
+        objs = self.read_output()
+        assert len(objs) == 1
+        o = objs[0]
+
+        assert o["type"] == "nfs"
+        assert o["rpc.auth_flavor"] == "unix"
+	assert "rpc.time" in o
```
Comment (andrewkroh) [inline packetbeat/tests/system/test_0061_nfs.py]: There appears to be a few tabs here that are throwing off the indentation. Please convert them to [spaces](http://legacy.python.org/dev/peps/pep-0008/#tabs-or-spaces). Thanks

Comment (kofemann) [inline packetbeat/tests/system/test_0061_nfs.py]: sorry about that. fixed.

**Comment by andrewkroh** [comment]
Looks good. I would like to get at least one more reviewer on this. @elastic/beats

**Comment by kofemann** [comment]
I will create a separate PR for dashboard later on.

File: packetbeat/etc/fields.yml [hunk 9]
```diff
@@ -1288,6 +1288,74 @@ trans_event:
           description: >
             The cursor identifier returned in the OP_REPLY. This must be the value that was returned from the database.
 
+    - name: rpc
```
Comment (andrewkroh) [inline packetbeat/etc/fields.yml]: Please run `make update` to re-generate the `packetbeat/docs/fields.asciidoc` document.

**Comment by monicasarbu** [comment]
@kofemann Thank you for the complete PR and the nice dashboard. I have only a small concern. 

I think it would be easier for the users to understand if we rename the Packetbeat module to `nfs` to be the same with the name of the configuration section `nfs`. In this case, we can move all the RPC information from `rpc` under `nfs.rpc`.

**Comment by kofemann** [comment]
@monicasarbu I had the same idea during prototyping was renaming it already several times :). The reason to call it rpc is that it will be easy to add other oncrpc/sunrpc based protocols as well. Nevertheless, I am fine to rename it. @andrewkroh any strong opinion?

**Comment by andrewkroh** [comment]
I would rename to `nfs`. If in the future new RPC based protocols are added we can refactor a bit to reuse this code.

**Comment by kofemann** [comment]
updated. The fields are still rpc.xxx and nfs.xxx. This makes more sense. You don't put **tcp** relates staff into **http** namspace

**Comment by monicasarbu** [comment]
@kofemann Ok, I agree with you that it makes more sense to have `rpc` and `nfs` inside the nfs event type.

**Comment by kofemann** [comment]
thanks! About dashboard: shold it go into packetbeat/etc/kibana ? Is there any conventions?

**Comment by monicasarbu** [comment]
@kofemann I am working now at a simple way to import and export the Kibana dashboard. For now, you can only export all the Kibana dashboards and dependencies (visualizations, searches, index-patterns) using the python script from https://github.com/elastic/beats-dashboards/tree/master/save. Please send us a PR against the beats-dashboards repo with all the exported dashboards and dependencies. I can help you to sort them out once the PR is opened. Thanks!

**Pull request merged**